blog for LongCat-Flash: Deploying Meituan's Agentic Model with SGLang (#193)
lm-sys/lm-sys.github.io: blog/2025-09-01-sglang-longcat-flash.md

![benchmark_overview.jpg](/images/blog/longcat_flash/benchmark_overview.jpg)

For more details, please refer to the our tech report [LongCat-Flash-Technical-Report](https://github.com/meituan-longcat/LongCat-Flash-Chat/blob/main/tech_report.pdf).

### 2. Why model-system co-design matters?

As noted in our tech report, a typical ReACT-based agent system imposes extreme requirements on both Prefill and Decode speeds due to its multi-round dialogue nature.

**For Prefill**, we observe that not every token in the model requires equal activation parameters. Based on this, we designed a dynamic activation feature that maintains model performance while controlling per-token activated parameters within 18.6B–31.3B (27B on average), which is crucial for reducing Prefill computation.

> **For decode**, high sparsity of MoE models necessitates large batches to increase GEMM (General Matrix Multiply) compute intensity. While wide Expert Parallelism (EP) frees up HBM for larger KV caches—effectively increasing batch size, communication becomes the bottleneck. Overlapping computation/communication is key to performance. TBO (Two Batch Overlap) from DeepSeekV3/SGLang reduces latency via batch overlap but fails for small batches or single request. Throughput (large batch) and latency (small batch) are inherently conflicting objectives and often need trade-off for online applications. Through model-system co-design, ScMoE breaks this trade-off by optimizing both throughput and latency. Another advantage from ScMoE is that intra-node Tensor Parallelism communication (via NVLink) on the dense FFN can be fully overlapped with inter-node Expert Parallelism communication (via RDMA), thereby maximizing total network utilization.

Reviewer: ```suggestion
**For decode**, high sparsity of MoE models necessitates large batches to increase GEMM (General Matrix Multiply) compute intensity. While wide Expert Parallelism (EP) frees up HBM for larger KV caches—effectively increasing batch size, communication becomes the bottleneck. Overlapping computation/communication is key to performance. TBO (Two Batch Overlap) from DeepSeek V3/SGLang reduces latency via batch overlap but fails for small batches or single request. Throughput (large batch) and latency (small batch) are inherently conflicting objectives and often need trade-off for online applications. Through model-system co-design, ScMoE breaks this trade-off by optimizing both throughput and latency. Another advantage from ScMoE is that intra-node Tensor Parallelism communication (via NVLink) on the dense FFN can be fully overlapped with inter-node Expert Parallelism communication (via RDMA), thereby maximizing total network utilization.
```

- **Latency-Optimized Scenarios**: SBO’s intra-batch optimization enables minimal latency.

**Benchmarks:**
- **Comparable throughput to DeepSeekV3**: Outperforms in generation speed.
- **Balanced throughput-latency**: Tested on the NVIDIA H800 platform, achieves **100 tps** with competitive cost.

| Model | Attention | Context | GPU | TGS | TPS/u |
| --- | --- | --- | --- | --- | --- |
| DeepSeek-V3-Profile | BF16 | 4096 | 128 | 2324 | 20 |
| LongCat-Flash | BF16 | 5000 | 128 | 2205 | 68.9 |
| LongCat-Flash | BF16 | 5000 | 128 | 804 | 100.5 |

### 5. Deploying LongCat-Flash with SGLang                                                            

We recommend deploying the LongCat-Flash using SGLang. Through close collaboration with the SGLang community, LongCat-Flash is supported on SGLang starting from day one. Due to its size of 560 billion parameters (560B), LongCat-Flash requires at least one node with 8xH20-141G node to host the model weights in FP8 format, and at least two nodes with 16xH800-80G for BF16 weights. Detailed launch configurations are provided below.

#### **Install SGLang**

```Shell
pip install --upgrade pip
pip install uv

> uv pip install "sglang[all]>=0.5.1.post3"

Reviewer: ```suggestion
uv pip install "sglang[all]>=0.5.2.rc0"
```

```
#### **Enabling Multi-Token Prediction (MTP)**

To enable MTP with SGLang, you can add the following arguments to your launch command.
```Shell
    --speculative-draft-model-path meituan-longcat/LongCat-Flash-Chat \
    --speculative-algorithm NEXTN \
    --speculative-num-draft-tokens 2 \
    --speculative-num-steps 1 \
    --speculative-eagle-topk 1
```
### 6. Conclusion

By leveraging capabilities such as SGLang, PD Disaggregation, Wide Expert Parallelism, and SBO, we have achieved extremely low cost and fast generation speed for LongCat-Flash. The efficient inference of LongCat-Flash also relies on works from the SGLang team, MoonCake team, NVIDIA trt-llm, and other open-source communities. Moving forward, we plan to collaborate with the SGLang team to gradually upstream our SGLang-based optimizations to further support the open-source ecosystem.

#### Acknowledgments

We would like to express our heartfelt gratitude to the following teams and collaborators:
- **SGLang Team and community:** for their work on SGLang framework.
- **Mooncake Team** for their earliest opensource work in the industry on PD Disaggregation architecture and TransferEngine.

> - **NVIDIA trt-llm:** for efficient kernels on Hopper GPUs.

Reviewer: ```suggestion
- **NVIDIA TensorRT-LLM:** for efficient kernels on Hopper GPUs.
```
